3664 Fix test_results_ordering failing test and other related fixes

albertisfu opened the pull request:
This PR fixes the randomly failing test **`test_results_ordering`**. The failure was due to sorting results by **`entry_date_filed`**, where a docket with a filing lacking a **`date_filed`** and a docket with no filings would receive the same score. Consequently, the order of these cases sometimes changed randomly.

The solution was not straightforward. Also during the investigation and testing, I noticed an additional issue related to party filtering, which I will describe later.

The sorting issue was as follows:

When ordering by **`entry_date_filed`**, we use a function score to compute the score of child documents based on the document field, in this case, the entry **`date_filed`**.

Depending on the order the score is computed, **`desc`** returns the **`date_filed`** in epoch time, boosting newer documents to appear first. If **`asc`**, the score returned is the difference **`current_time - date_filed`**, showing older documents first.

However, if a document lacks a **`date_filed`** (None), it returned a score of 0.

The maximum score returned by the inner hits is used as the hit score, allowing cases to be sorted by the child field **`entry_date_filed`**.

The problem arises when a docket without filings can have the same or a greater score than a case with a filing but no **`date_filed`**.

The correct logic is that if the user wishes to order by **`entry_date_filed`**, it is expected that cases with dockets are shown first, and dockets without documents that also match the query should come last.

However, as explained before, a case with a document but no **`date_filed`** was returning 0, while a docket without documents could have a score higher than 0 because, for dockets without filings, the function score for **`has_child`** documents is not executed.

The solution was to return a score of 1 for dockets with a filing but no **`date_filed`**, instead of 0. Additionally, the method **`nullify_parent_score_on_child_sorting`** was introduced. This method sets the score returned by the parent query to 0, ensuring that when ordering by a child field like **`entry_date_filed`**, the parent query score is always 0. Thus, the score computed in the **`has_child`** query is the only score in the result, while dockets without filings remain at 0, ensuring they always come after all cases with filings, regardless of their score.

This solution worked for all test cases except one: when filtering cases without filings by a party filter.

In issue #3641, we resolved a problem related to party filters. Essentially, a filter that involved a party field and any other field would include cases that did not match the parties at all.

This issue was fixed, and now, when filtering by parties and other fields, the filters work correctly.

However, I overlooked a test case where filtering by party fields excluded dockets without filings, even if the docket matched the parties. This is because the new query applied only when a party field is used is as follows:

```
bool
  should
     has_child_query
          filters_without_parties
          query_string
     parent_query
         filters_including_parties
         query_string
 filter
    filters_including_parties
    has_child_query
        filters_without_parties
        query_string
```

The new filter level on party queries:
```
 filter
    filters_including_parties
    has_child_query
        filters_without_parties
        query_string
```
Uses the `has_child_query` as a filter so it'll only return matches with child documents, excluding dockets without filings even if they matched the parties filter.

So the solution was to tweak the query a bit more:
```
bool
  should
     has_child_query
          filters_without_parties
          query_string
     parent_query
         filters_including_parties
         query_string
must
   should
        must
            filters_including_parties
            has_child_query
                filters_without_parties
                query_string
        must
            filters_including_parties
            only_dockets
```

So now, instead of using **`filter`**, we're employing **`must`**. This change functions similarly, but **`must`** allows for score control via a function score. Additionally, it includes a **`should`** clause that can match either dockets with filings that matched the **`has_child_query`** or dockets without filings that matched the parties filter.

This resolves the issue of not being able to match dockets without filings when using the party filters.

However, there remains a scenario where it is not possible to match dockets without filings: if you combine the party filters with any other filter or a query string, it will only match cases with filings.

I tried different approaches, but none of them worked. This seems to be part of the trade-off when we decided to index only parties in the `DocketDocument`.

When combining party filters with any other filter or string query, there are two possible outcomes:

- The issue we previously addressed and resolved in #3641, where a query involving party filters and any other filter can match dockets with filings that meet the query criteria except for the party filters, plus dockets that only match the party filters. Here, users perceived a massive amount of results so they think the party filter does not work.
- The issue described in the previous paragraphs. Combining party filters with any other filter or string query will match only dockets with filings, excluding empty dockets. The only way to match an empty docket by parties is to exclusively use party filters (party and/or attorney) without any other filters.

I believe the last approach, which is implemented in this PR, has the best trade-off. Users will be able to use party filters alongside any other filter or query string for all dockets with filings. Additionally, they can still filter by parties and match dockets without filings if they exclusively use the party filters. This can be clarified further in the documentation.

Additionally, I attempted to trigger the **`test_new_recap_email_alerts_integration`** failing test but couldn't reproduce it. However, upon a detailed examination of the process, I believe the problem might be due to a race condition with another test that enqueues a docket alert for the same docket ID. Therefore, I added a mock to **`enqueue_docket`** in all recap.email-related tests, hoping this resolves the failing test issue.

I also introduced a mock in **`test_are_events_created_properly`**, a test that Eduardo noticed was also failing randomly. He suggested adding a mock for **`cl.api.utils.get_logging_prefix`** to prevent a race condition with another test running in parallel from affecting this test.

mlissner commented on cl/search/tests/tests_es_recap.py:
```diff
@@ -1068,9 +1081,12 @@ def test_party_name_and_children_filter(self) -> None:
 
         ## The party_name and attorney filter can constrain the results
         # returned, along with string query, parent and child filters.
+        # Note that the case 'California v. America' is not found here.
+        # Dockets without filings cannot be located by parties fields if
+        # combined with any other filter or string query.
```
Would this work better as an assertion with a comment?

albertisfu replied:
Sure I've added the assertion and the comment.
- Also added an additional test to confirm that it is possible to match a docket without filings, as in this case, by using a fielded search and a docket-level field filter.
e.g:
```
"q": attorney:(Harris Martin)",
"case_name": "California",
```

mlissner reviewed (approved):
Thanks for the detailed analysis. This all looks as far as I can tell, but probably good to have somebody else look carefully at it too, so I'll put this on Eduardo's backlog to review too.

A few thoughts:

1. Any idea why the tests started failing?

3. Will you take a stab at a simple explanation in the docs either in this PR or another one?

Thanks for figuring this out!

albertisfu commented:
Thanks for your comments!

> Any idea why the tests started failing?

My theory is that the test could have failed since it was introduced. However, after adding more tests, the order of documents in the shard might have changed. Since the documents in the test that was failing had the same score, sometimes they were retrieved in the expected order, and other times their order switched.

> Will you take a stab at a simple explanation in the docs either in this PR or another one?

Sure, I've created https://github.com/freelawproject/courtlistener/issues/3685 for it where I put some questions there regarding the section where this explanation should go.

mlissner commented:
Great. All yours, Eduardo.

ERosendo commented on cl/lib/elasticsearch_utils.py:
```diff
@@ -963,10 +975,18 @@ def build_child_docs_query(
+    if search_query and must_query:
+        # Extract parent filters from the query, if any, for building the
+        # has_parent query.
+        try:
+            query_strings = search_query.query.must[0].should[0].must
+        except (AttributeError, IndexError):
+            query_strings = must_query
```
We should use the `build_join_es_filters` method instead of extracting the query strings from the search query since the ES schema is not always the same.

albertisfu replied:
Sure, I applied this suggestion, thanks!

ERosendo reviewed (approved):
The code looks good. We just need to address the suggestion in my previous comment, and then we'll be ready to merge this PR.

ERosendo reviewed (approved):
The refactor is working properly. I tried the following combinations to test the new party query: 

Scenarios:

- Party Name filter only:
   - Retrieves dockets with documents and empty dockets with matching parties.
   
- Party Name filter + Case Name filter(docket filter):
   - Retrieves dockets meeting both criteria, including empty ones.

- Party Name filter + Case Name filter(docket filter) + Document Description filter(doc filter):
   - Successfully narrows results to dockets containing documents that align with all three filters.

albertisfu commented:
@mlissner You were right about your comment in #3688. The issue was related to matching parties and other docket fields. So, it was incorrect to say that it wouldn't be possible to filter cases by parties and other parent fields, since both belong to Dockets.

> If a case has no filings, how would a non-docket field do anything anyway?

In this case, I think it's correct that if you use a child field as a filter, it shouldn't match empty dockets. This is because the child field constrains the query to match only documents with filings.

I've refactored the RECAP Search query once more, simplifying it by removing the additional level of filters that was previously used for party filters. The new query structure when a party field is in use is as follows:

```
bool
  should
    bool
       must
         parties_filters
         has_child_query
            filters_without_parties
            query_string
     parent_query
         filters_including_parties
         query_string
```

I've added additional test cases to confirm the following capabilities of the new approach:
- Matching empty dockets: It can correctly match empty dockets when combining party filters with other parent filters.
- Excluding empty dockets: When a query includes both a party filter and a child filter, it appropriately excludes empty dockets from the results.
